### PR TITLE
[FEAT]: Accept invitation to join organization -Backend

### PR DESCRIPTION
**Pull request by Olatomiw**
<h1> How should this be manually tested </h1>

- This feature can be tested on Postman manually by copying and pasting the invitation link in the request body of "api/v1/organisation/invite" 
```
{"invitationRequest" : "http://api/hello?organisationId=8469de43-1e63-4bf9-b802-424ac84cb60d&userId=0309520e-499a-407b-8baf-3ab2c0e805ea&expires=2025-07-23T00:00:00Z"} 
```

- The link is in that form and you will have to manually update the request params in the link as there's no one currently working on the invitation link. So i just had to come up with this approach
- This is a protected endpoint. So you also need your token before you can access it\

<h1>Checklist of what I did</h1>

- [x] Validate the the invitation Link. i.e check if the organisation exist and the invitation link has not expired yet
- [x] I also check if user doesn't already belong to the organization. To avoid duplicate
- [x]  Accept the appropriate user into the desired organization
- [x] Unit testing to check if the link has expired, the link is malformed and finally accepts the user into the organisation

## Reference of Issue
[https://github.com/hngprojects/hng_boilerplate_java_web/issues/4#issue-2416092868](ok)

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/organisation/controller/OrganisationController.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+@RestController
+@RequestMapping("/api/v1/org")
+public class OrganisationController {
+
+    @PostMapping("/invite")
+    public ResponseEntity<?> acceptInvitation(@RequestBody String invitationLink){
+        if (invitationLink == " " || invitationLink == null){
+
+        }
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/controller/OrganisationController.java:24]: Remove empty space

File: src/main/java/hng_java_boilerplate/organisation/controller/OrganisationController.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        return null;
+    }
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/controller/OrganisationController.java:29]: Remove empty space

File: src/main/java/hng_java_boilerplate/organisation/dto/InvitationRequest.java
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/dto/InvitationRequest.java:1]: If it's not needed delete it

File: src/main/java/hng_java_boilerplate/organisation/exception/InvitationValidationException.java
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/exception/InvitationValidationException.java:1]: If it's not needed delete it

File: src/main/java/hng_java_boilerplate/organisation/exception/OrganisationException.java
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/exception/OrganisationException.java:1]: If it's not needed delete it

File: src/main/java/hng_java_boilerplate/organisation/exception/exceptionHandler/MyExceptionHandler.java
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/exception/exceptionHandler/MyExceptionHandler.java:1]: If it's not needed delete it

File: src/main/java/hng_java_boilerplate/organisation/exception/response/ErrorResponse.java
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/exception/response/ErrorResponse.java:1]: If it's not needed delete it

File: src/main/java/hng_java_boilerplate/organisation/exception/response/SuccessResponse.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package hng_java_boilerplate.organisation.exception.response;
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/exception/response/SuccessResponse.java:2]: Remove space

File: src/main/java/hng_java_boilerplate/organisation/exception/response/SuccessResponse.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+import lombok.AllArgsConstructor;
+import org.springframework.http.HttpStatus;
+@AllArgsConstructor
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/exception/response/SuccessResponse.java:5]: Add a space

No data annotation

File: src/main/java/hng_java_boilerplate/organisation/service/OrganisationService.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    public Organisation getOrganisationDetails(String orgId) throws OrganisationException{
+        Organisation foundOrganisation = organisationRepository.findById(orgId)
+                .orElseThrow(()-> new OrganisationException("Invalid Organisation Id"));
+        return foundOrganisation;
+    }
+
+    public SuccessResponse validateAndAcceptUserToOrganisation(String OrgId){
+//        This is a secured method and while i await Auth guy to come through so i can fetch the id of logged-in user
+// TODO: 7/20/2024 fetch user that wants to be added to organisation from the logged in user
+//        A demo user
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/service/OrganisationService.java]: When you signup a user the user is logged in, use that for now

hard coding a user id won't fly

**Review by Am0du — CHANGES_REQUESTED** [review]
Add Unit test

File: src/main/java/hng_java_boilerplate/organisation/service/OrganisationService.java (hunk 5 above)
Comment (Olatomiw) [reply src/main/java/hng_java_boilerplate/organisation/service/OrganisationService.java]: please sir. I don't think you checked the latest commit . All these you have mentioned are no longer on my code

File: src/main/java/hng_java_boilerplate/organisation/dto/InvitationRequest.java [hunk 6]
```diff
@@ -14,14 +14,14 @@ public class InvitationRequest {
     @Autowired
     private OrganisationService organisationService;
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/organisation/dto/InvitationRequest.java:15]: Avoid field injection, also do you need the service in a dto?
Use data annotation

**Review by Am0du — COMMENTED** [review]
checkout these changes and you're good to go

**Review by Am0du — APPROVED** [review]

**Pull request closed without merging**